src/Compilers/CSharp/Portable/BoundTree/Expression.cs

```diff
@@ -100,103 +78,120 @@ public override void Accept(OperationVisitor visitor)
-                {
-                    Symbols.ParameterSymbol lastParameter = parameters[parameters.Length - 1];
-                    if (lastParameter.IsParams)
+                ConcurrentDictionary<Symbols.ParameterSymbol, IArgument> omittedArguments = s_omittedArgumentMappings.GetValue(invocationSyntax, syntax => new ConcurrentDictionary<Symbols.ParameterSymbol, IArgument>());
+
+                return omittedArguments.GetOrAdd(
+                    parameters[parameterIndex],
+                    (parameter) =>
                     {
-                        return new Argument(ArgumentKind.ParamArray, lastParameter, CreateParamArray(lastParameter, boundArguments, argumentIndex, invocationSyntax));
-                    }
-                }
+                        // No argument has been supplied for the parameter at `parameterIndex`:
+                        // 1. `argumentIndex == -1' when the arguments are specified out of parameter order, and no argument is provided for the parameter corresponding to `parameters[parameterIndex]`.
+                        // 2. `argumentIndex >= boundArguments.Length` when the arguments are specified in parameter order, and no argument is provided at `parameterIndex`.
+
+                        // Check for a parameter with a default value.
+                        if (parameter.HasExplicitDefaultValue)
+                        {
+                            return new Argument(ArgumentKind.DefaultValue, parameter, new LiteralExpression(parameter.ExplicitDefaultConstantValue, parameter.Type, invocationSyntax));
+                        }
+
+                        // Check for an omitted argument that becomes an empty params array.
+                        if (parameter.IsParams)
+                        {
+                            return new Argument(ArgumentKind.ParamArray, parameter, CreateParamArray(parameter, boundArguments, argumentIndex, invocationSyntax));
+                        }
 
-                // There is no supplied argument and there is no params parameter. Any action is suspect at this point.
-                return new SimpleArgument(null, new InvalidExpression(invocationSyntax));
+                        // There is no supplied argument and there is no params parameter. Any action is suspect at this point.
+                        return new Argument(ArgumentKind.DefaultValue, parameter, new InvalidExpression(invocationSyntax));
+                    });
             }
 
             return s_argumentMappings.GetValue(
                 boundArguments[argumentIndex],
                 (argument) =>
                 {
-                    string nameOpt = !argumentNamesOpt.IsDefaultOrEmpty ? argumentNamesOpt[argumentIndex] : null;
-                    Symbols.ParameterSymbol parameterOpt = (uint)parameterIndex < (uint)parameters.Length ? parameters[parameterIndex] : null;
+                    string name = !argumentNames.IsDefaultOrEmpty ? argumentNames[argumentIndex] : null;
```

Comment:
is argumentNames never default?  If so, the name change is ok.  If it can be default, the name should stay the same.  If it can't be default, then no need to call .IsDefault.  

it looks like name can be null.  please call it nameOpt.